File: src/main/java/com/darkyen/paragrowth/doodad/DoodadShader.java
package com.darkyen.paragrowth.doodad;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.g3d.Attributes;
import com.badlogic.gdx.graphics.g3d.Renderable;
import com.badlogic.gdx.graphics.g3d.Shader;
import com.badlogic.gdx.graphics.g3d.shaders.BaseShader;
import com.badlogic.gdx.graphics.g3d.shaders.DefaultShader;
import com.darkyen.paragrowth.util.PrioritizedShader;
import com.darkyen.paragrowth.util.AutoReloadShaderProgram;

/**
 *
 */
class DoodadShader extends BaseShader implements PrioritizedShader {

    private DoodadShader() {
        register(DefaultShader.Inputs.projViewWorldTrans, DefaultShader.Setters.projViewWorldTrans);
    }

    @Override
    public void init() {
        throw new UnsupportedOperationException("Do not init me directly!");
    }

    @Override
    public int compareTo(Shader other) {
        if (other == null) return -1;
        if (other == this) return 0;
        return 0;
    }

    @Override
    public boolean canRender(Renderable instance) {
        return true;
    }

    @Override
    public void render (Renderable renderable, Attributes combinedAttributes) {
        context.setBlending(false, GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA);
        context.setCullFace(GL20.GL_NONE);
        context.setDepthTest(GL20.GL_LESS);
        context.setDepthMask(true);

        super.render(renderable, combinedAttributes);
    }

    private static DoodadShader INSTANCE = null;
    static DoodadShader get(Renderable renderable) {
        DoodadShader instance = INSTANCE;
        if (instance != null) {
            return instance;
        }

        instance = new DoodadShader();
        instance.init(new AutoReloadShaderProgram(Gdx.files.local("doodad_vert.glsl"), Gdx.files.local("doodad_frag.glsl")), renderable);

        INSTANCE = instance;
        return instance;
    }

    @Override
    public int priority() {
        return DOODADS;
    }
}

File: src/main/java/com/darkyen/paragrowth/doodad/DoodadShader.kt
package com.darkyen.paragrowth.doodad

import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.GL30
import com.badlogic.gdx.graphics.g3d.utils.RenderContext
import com.darkyen.paragrowth.render.*
import com.darkyen.paragrowth.util.WORLD_BLEND_ATTRIBUTE
import com.darkyen.paragrowth.util.setupBlendWallUniforms

val VA_BLEND_OFFSET = VertexAttribute("a_blend_offset", GL30.GL_FLOAT, 1)

val DOODAD_ATTRIBUTES = VertexAttributes(
        VA_POSITION3,
        VA_COLOR1,
        VA_BLEND_OFFSET
)

val DOODAD_SHADER_BLEND_IN = DoodadShader(true)
val DOODAD_SHADER_BLEND_OUT = DoodadShader(false)

/**
 *
 */
class DoodadShader(blendIn:Boolean) : Shader(DOODADS, "doodad", DOODAD_ATTRIBUTES) {

    init {
        globalUniform("u_projViewTrans") { uniform, camera, _ ->
            uniform.set(camera.combined)
        }

        globalUniform("u_blend") { uniform, _, attributes ->
            uniform.set(attributes[WORLD_BLEND_ATTRIBUTE][0])
        }

        globalUniform("u_blendIn") { uniform, camera, attributes ->
            uniform.set(if(blendIn) 1 else 0)
        }

        setupBlendWallUniforms()
    }

    override fun adjustContext(context: RenderContext) {
        context.setBlending(false, GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA)
        context.setCullFace(GL20.GL_NONE)// TODO(jp): Enable, after fixing winding
        context.setDepthTest(GL20.GL_LESS)
        context.setDepthMask(true)
    }
}